newrelic-agent/src/test/java/com/newrelic/agent/service/logging/LogSenderServiceImplTest.java

```diff
@@ -1,7 +1,234 @@
+        logSenderService.recordLogEvent(ImmutableMap.of("field3", "value3"));
+
+        MockRPMService analyticsData = new MockRPMService();
+        when(ServiceFactory.getServiceManager().getRPMServiceManager().getOrCreateRPMService(appName)).thenReturn(
+                analyticsData);
+
+        logSenderService.harvestHarvestables();
+
+        assertEquals(3, analyticsData.getEvents().size());
+
+        logSenderService.stop();
+
+        verify(txService, times(1)).removeTransactionListener(logSenderService.transactionListener);
+    }
+
+    @Test
+    public void testWithTransaction() throws Exception {
+        LogSenderServiceImpl logSenderService = createService(createConfig(null, 180));
+        Transaction transaction = Mockito.mock(Transaction.class);
+        when(ServiceFactory.getTransactionService().getTransaction(false)).thenReturn(transaction);
+
+        LogSenderServiceImpl.TransactionLogs logs = new LogSenderServiceImpl.TransactionLogs(
+                AgentConfigImpl.createAgentConfig(Collections.emptyMap()));
+        when(transaction.getLogEventData()).thenReturn(logs);
+        when(transaction.getApplicationName()).thenReturn(appName);
+        when(transaction.isInProgress()).thenReturn(true);
+
+        logSenderService.recordLogEvent(ImmutableMap.of("field", "value"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field2", "value2"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field3", "value3"));
+
+        MockRPMService analyticsData = new MockRPMService();
+        when(ServiceFactory.getServiceManager().getRPMServiceManager().getOrCreateRPMService(appName)).thenReturn(
+                analyticsData);
+
+        logSenderService.harvestHarvestables();
+
+        assertEquals(0, analyticsData.getEvents().size());
+        assertEquals(3, logs.events.size());
+    }
+
+    @Test
+    public void testTransactionHarvest() throws Exception {
+        LogSenderServiceImpl logSenderService = createService(createConfig(null, 180));
+        logSenderService.addHarvestableToService(appName);
+
+        Transaction transaction = Mockito.mock(Transaction.class);
+        when(ServiceFactory.getTransactionService().getTransaction(false)).thenReturn(transaction);
+
+        LogSenderServiceImpl.TransactionLogs logs = new LogSenderServiceImpl.TransactionLogs(
+                AgentConfigImpl.createAgentConfig(Collections.emptyMap()));
+        when(transaction.getLogEventData()).thenReturn(logs);
+        when(transaction.getApplicationName()).thenReturn(appName);
+        when(transaction.isInProgress()).thenReturn(true);
+
+        logSenderService.recordLogEvent(ImmutableMap.of("field", "value"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field2", "value2"));
+        logSenderService.recordLogEvent(ImmutableMap.of("field3", "value3"));
+        // TODO is this valid?  should an event with empty map of attributes be included?
```

Comment:
How do I create invalid log event that will be filtered out?

Reply:
Validation logic is here: https://github.com/newrelic/newrelic-java-agent/blob/logging-initiative/newrelic-agent/src/main/java/com/newrelic/agent/service/logging/LogSenderServiceImpl.java#L468-L510

It's mostly validating the attributes on the LogEvent. Look for things that should be filtered out like like `null` for attribute key/value, unsupported attribute type (`Each key should be a String and each value should be a String, Number, or Boolean`), etc.

Reply:
Ok, I'll try this.  I tried an empty map of attributes and it took it.  I'll try some other combos.

Reply:
Ok, so it looks like at the point where it gets called from `AgentUtil` the attribute map will always have at least two entries.  So no way an event can currently be logged with null or empty map of attributes anyways.

Reply:
What about cases where key or value is invalid? I don't see that covered in the tests.

```
logSenderService.recordLogEvent(ImmutableMap.of("field", null));
logSenderService.recordLogEvent(ImmutableMap.of(null, "value"));
```

Reply:
I've done them, but it does not cause the entire event to get removed, just those attributes.

Reply:
> I've done them, but it does not cause the entire event to get removed, just those attributes.

That's the expected behavior. Assertions should be on only the valid attributes being present.

Reply:
Ok, let's chat.  Worked with @meiao and @tbradellis to add these tests and fix the code.  The issue was that although the invalid entries were filtered, it could lead to logging an event with empty map of attributes which based on other code is not valid.